### ISSUE TITLE
Enhancement: strict-boolean-expressions -- option to not error on always truthy objects

### ISSUE THREAD

**Issue by stubailo** [typescript-eslint/typescript-eslint#8175]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/strict-boolean-expressions/#examples

### Description

I'd like to use the `strict-boolean-expressions` rule to only check for numbers -- so essentially set `allowNullableNumber: false, allowNumber: false`, but set all the rest to true.

That's great, but there is a not-configurable condition here that always reports an error on a truthy object:

https://github.com/typescript-eslint/typescript-eslint/blob/e566a5dda347470b8ced3cc301b7e4d3e7ed721b/packages/eslint-plugin/src/rules/strict-boolean-expressions.ts#L698-L703

It would be great to have an option `allowObject: true` that turns off this error.

### Situation

We have [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) turned off in our tsconfig. It would be a lot of work to enable it at this type, so we're trying to incrementally improve types.

This means that if you have an array of objects, or an object of objects, then it's easy to end up with objects that appear to be always defined, but can be unexpectedly undefined.

### Fail

```TypeScript
// N/A
```

### Pass

```TypeScript
// In this example, we have a list of database IDs and we want to get an object of arrays,
// where the keys are the foreign keys and the values are arrays of objects.
// We want to ensure the object contains a key for every ID, even if no items were returned.
const userIds: number[] = [1, 2, 3];

const posts = db.Posts.getByUserIds(userIds);
const postsByUserId = _.groupBy(posts, 'userId'); // Returns a Record<number, Post> basically

for (const userId of userIds) {
  if (!postsByUserId[userId]) { // This is the line where the linter would error
    postsByUserId[userId] = [];
  }
}

return postsByUserId;
```

### Additional Info

I see that https://github.com/typescript-eslint/typescript-eslint/issues/7703 which is similar was closed, but I believe that this is a valid use case where it's hard to make the TS types actually match the reality without turning on `noUncheckedIndexedAccess`, which is a pretty heavy-duty rule.

**Comment by damacisaac** [comment]
Same request here. Trying to enable the rule to only check for numbers, which in our case is the only scenario where we've had bugs. This would require:

1. an option for `allowObject`
2. an option for `allowTruthy`
3. an option to disable the `conditionErrorOther` report

**Comment by bradzacher** [comment]
7 months later and this request has received 0 community reactions.
As per our [workflow guidelines](https://typescript-eslint.io/maintenance/issues#community-engagement-evaluation) - we are taking this as signal that this check is not important to the broader community and are thus rejecting the proposal. Thanks for filing!

**Comment by stubailo** [comment]
Would you accept a PR for this option?

**Comment by bradzacher** [comment]
No - as per my comment and the linked guidelines this addition has been requested for inclusion in our project on the basis that it fails the following check

> I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).
